In [1]:
! pip install selenium

In [2]:
# Importing Libraries
import selenium
import pandas as pd
import time
from bs4 import BeautifulSoup

# Importing selenium webdriver 
from selenium import webdriver

# Importing required Exceptions which needs to handled
from selenium.common.exceptions import StaleElementReferenceException, NoSuchElementException,ElementNotInteractableException,ElementNotVisibleException

#Importing requests
import requests

# importing regex
import re

In [3]:
# Activating the chrome browser
driver=webdriver.Chrome("chromedriver.exe") 
time.sleep(3)

In [6]:
# opening the homepage of cars24
url="https://www.cars24.com/buy-used-cars-new-delhi/?itm_source=Cars24Website&itm_medium=sticky_header"
driver.get(url)

In [7]:
# Here the location is New Delhi

In [8]:
#scrapping name of the car
car_name=[]
for page in range(1,139):
    try:
        name=driver.find_elements_by_xpath('//h2[@class="_3FpCg"]')
        for i in name:
            car_name.append(i.text)
    except NoSuchElementnException:
        car_name.append("No details available")

In [9]:
len(car_name)

2760

In [10]:
car_model=[]
for page in range(1,139):
    try:
        model=driver.find_elements_by_xpath('//p[@class="cvakB"]')
        for j in model:
            car_model.append(j.text.replace("Manual",""))
    except NoSuchElementnException:
        car_model.append("No details available")

In [11]:
len(car_model)

2760

In [12]:
car_price=[]
for page in range (1,139):
    try:
        price=driver.find_elements_by_xpath("//p[@class='_31Wtw']")
        for k in price:
            car_price.append(k.text.replace("₹",""))
    except NoSuchElementException:
        car_price.append("No details available")

In [13]:
len(car_price)

2760

In [14]:
trans=[]
for page in range(1,139):
    try:
        tran=driver.find_elements_by_xpath("//p[@class='cvakB']/span")
        for l in tran:
            trans.append(l.text)
    except NoSuchElementException:#handling no such element exception
        trans.append('No details available')

In [15]:
len(trans)

2760

In [18]:
km=[]
for page in range (1,139):
    try:
        kms=driver.find_elements_by_xpath("//ul[@class='bVR0c']/li[1]")
        for m in kms:
            km.append(m.text.replace("km",""))
    except NoSuchElementException:#handling no such element exception
        km.append('No details available')
    except StaleElementReferenceException:#handling Stale element exception
        km.append('No details available')

In [21]:
len(km)

2760

In [25]:
owner=[]
for page in range(1,139):
    try:
        own=driver.find_elements_by_xpath("//ul[@class='bVR0c']/li[2]")
        for n in own:
            owner.append(n.text)
    except NoSuchElementException:#handling no such element exception
        owner.append('No details available')
    except StaleElementReferenceException:#handling Stale element exception
        owner.append('No details available')

In [26]:
len(owner)

2760

In [27]:
fuel=[]
for page in range(1,139):
    try:
        t_fuel=driver.find_elements_by_xpath("//ul[@class='bVR0c']/li[3]")
        for o in t_fuel:
            fuel.append(o.text)
    except NoSuchElementException:#handling no such element exception
        owner.append('No details available')
    except StaleElementReferenceException:#handling Stale element exception
        owner.append('No details available')

In [29]:
len(fuel)

2760

In [49]:
discount=[]
for page in range(1,139):
    try:
        dis=driver.find_elements_by_xpath("//div[@class='_3mZ8j']")
        for p in dis:
            discount.append(p.text.replace("OFF",""))
    except NoSuchElementException:#handling no such element exception
        discount.append('No details available')
        

In [51]:
len(discount)

2760

In [52]:
dis_price=[]
for page in range(1,139):
    try:
        d_price=driver.find_elements_by_xpath("//div[@class='_7udZZ']")
        for q in d_price:
            dis_price.append(q.text.replace("₹",""))
    except NoSuchElementException:#handling no such element exception
        dis_price.append('No details available')
        

In [54]:
len(dis_price)

2760

In [56]:
ds=pd.DataFrame({"Name":car_name,
                 "Model":car_model,
                "Transmission":trans,
                "Km_driven":km,
                "Owner":owner,
                "Fuel":fuel,
                "Discount":discount,
                "Discount_price":dis_price,
                "Sale_price":car_price})
ds

Name                     Model Transmission  \
0           2020 Renault Kwid  1.0 RXT Opt AT Automatic    Automatic   
1        2013 Maruti Alto 800                LXI Manual       Manual   
2     2020 Maruti New Wagon-R            VXI 1.0 Manual       Manual   
3           2019 Maruti Swift                VXI Manual       Manual   
4          2020 Maruti Baleno      SIGMA 1.2 K12 Manual       Manual   
...                       ...                       ...          ...   
2755  2019 Maruti New Wagon-R            VXI 1.0 Manual       Manual   
2756        2018 Maruti Swift                LXI Manual       Manual   
2757      2014 Maruti Celerio         VXI AMT Automatic    Automatic   
2758  2013 Maruti Swift Dzire          VDI BS IV Manual       Manual   
2759        2017 Renault Kwid            CLIMBER 1.0 AT                

     Km_driven      Owner    Fuel  Discount Discount_price Sale_price  
0       3,252   1st Owner  Petrol   ₹5,000        4,31,699   4,36,699  
1      13,807   1st Owner  Petrol  ₹11,000        2,16,899   2,27,899  
2       3,865   1st Owner  Petrol   ₹5,000        5,16,399   5,21,399  
3       3,362   1st Owner  Petrol   ₹5,000        6,20,699   6,25,699  
4       5,645   1st Owner  Petrol   ₹6,000        5,75,299   5,81,299  
...        ...        ...     ...       ...            ...        ...  
2755   11,213   1st Owner  Petrol  ₹37,000        4,68,599   5,05,599  
2756    4,482   1st Owner  Petrol   ₹5,000        5,23,299   5,28,299  
2757   19,045   2nd Owner  Petrol  ₹16,000        3,25,399   3,41,399  
2758   33,554   1st Owner  Diesel   ₹5,000        3,70,199   3,75,199  
2759   11,721   1st Owner  Petrol  ₹15,000        3,83,799   3,98,799  

[2760 rows x 9 columns]

In [57]:
#Now the location is Chennai

In [125]:
# Activating the chrome browser
driver=webdriver.Chrome("chromedriver.exe") 
time.sleep(3)

In [126]:
# opening the homepage of cars24
url="https://www.cars24.com/buy-used-car?sort=P&storeCityId=5732&pinId=600001"
driver.get(url)

In [133]:
#scrapping name of the car
car_name=[]
car_model=[]
trans=[]
km=[]
owner=[]
fuel=[]
discount=[]
dis_price=[]
car_price=[]


for page in range(1,130):
    try:
        name=driver.find_elements_by_xpath('//h2[@class="_3FpCg"]')
        for a in name:
            car_name.append(a.text)
    except NoSuchElementnException:
        car_name.append("No details available")
        
time.sleep(2)
for page in range(1,130):
    try:
        model=driver.find_elements_by_xpath('//p[@class="cvakB"]')
        for b in model:
            car_model.append(b.text.replace("Manual",""))
    except NoSuchElementnException:
        car_model.append("No details available") 
        
time.sleep(2)  
for page in range(1,130):
    try:
        price=driver.find_elements_by_xpath("//p[@class='_31Wtw']")
        for c in price:
            car_price.append(c.text.replace("₹",""))
    except NoSuchElementException:
        car_price.append("No details available") 
        
time.sleep(2) 
for page in range(1,130):
    try:
        tran=driver.find_elements_by_xpath("//p[@class='cvakB']/span")
        for d in tran:
            trans.append(d.text)
    except NoSuchElementException:#handling no such element exception
        trans.append('No details available')  
        
time.sleep(2)

for page in range(1,130):
    try:
        kms=driver.find_elements_by_xpath("//ul[@class='bVR0c']/li[1]")
        for e in kms:
            km.append(e.text.replace("km",""))
    except NoSuchElementException:#handling no such element exception
        km.append('No details available')
    except StaleElementReferenceException:#handling Stale element exception
        km.append('No details available')
        
time.sleep(2) 

for page in range(1,130):
    try:
        own=driver.find_elements_by_xpath("//ul[@class='bVR0c']/li[2]")
        for f in own:
            owner.append(f.text)
    except NoSuchElementException:#handling no such element exception
        owner.append('No details available')
    except StaleElementReferenceException:#handling Stale element exception
        owner.append('No details available')

time.sleep(2)   
for page in range(1,130):
    try:
        t_fuel=driver.find_elements_by_xpath("//ul[@class='bVR0c']/li[3]")
        for g in t_fuel:
            fuel.append(g.text)
    except NoSuchElementException:#handling no such element exception
        owner.append('No details available')
    except StaleElementReferenceException:#handling Stale element exception
        owner.append('No details available')
        
        
time.sleep(2) 

for page in range(1,130):
    try:
        dis=driver.find_elements_by_xpath("//div[@class='_3mZ8j']")
        for h in dis:
            discount.append(h.text.replace("OFF",""))
    except NoSuchElementException:#handling no such element exception
        discount.append('No details available')
        
time.sleep(2) 

for page in range(1,130):
    try:
        d_price=driver.find_elements_by_xpath("//div[@class='_7udZZ']")
        for t in d_price:
            dis_price.append(t.text.replace("₹",""))
    except NoSuchElementException:#handling no such element exception
        dis_price.append('No details available')

# creating the dataframe
dt=pd.DataFrame({"Name":car_name,
                 "Model":car_model,
                "Transmission":trans,
                "Km_driven":km,
                "Owner":owner,
                "Fuel":fuel,
                "Discount":discount,
                "Discount_price":dis_price,
                "Sale_price":car_price})
dt

Name                   Model Transmission Km_driven  \
0            2020 Maruti Alto                    LXI        Manual    3,494    
1           2014 Maruti Swift                    VXI        Manual    3,600    
2            2011 Hyundai i10       MAGNA 1.2 KAPPA2        Manual   13,179    
3            2011 Maruti Alto                    LXI        Manual   23,217    
4            2014 Maruti Eeco  5 STR WITH AC PLUSHTR        Manual   18,371    
...                       ...                     ...          ...       ...   
2575  2016 Maruti Wagon R 1.0                    VXI        Manual   33,463    
2576         2015 Hyundai Eon                 SPORTZ        Manual   15,346    
2577  2012 Maruti Wagon R 1.0                    VXI        Manual   38,553    
2578          2018 Tata Tiago        XT 1.2 REVOTRON        Manual   11,253    
2579     2017 Maruti Alto 800                    VXI        Manual   21,239    

          Owner    Fuel  Discount Discount_price Sale_price  
0     1st Owner  Petrol   ₹5,000        4,19,599   4,24,599  
1     1st Owner  Petrol  ₹20,000        4,48,999   4,68,999  
2     1st Owner  Petrol   ₹8,000        3,18,999   3,26,999  
3     1st Owner  Petrol   ₹7,000        2,48,099   2,55,099  
4     1st Owner  Petrol  ₹10,000        3,54,499   3,64,499  
...         ...     ...       ...            ...        ...  
2575  1st Owner  Petrol   ₹6,000        3,88,699   3,94,699  
2576  3rd Owner  Petrol  ₹20,000        3,01,299   3,21,299  
2577  1st Owner  Petrol  ₹12,000        3,15,199   3,27,199  
2578  1st Owner  Petrol   ₹7,000        4,86,799   4,93,799  
2579  1st Owner  Petrol  ₹37,000        3,20,999   3,57,999  

[2580 rows x 9 columns]

In [139]:
# Let's concatenate bothe the dataframe ds and dt

In [140]:
df=pd.concat([ds,dt],axis=0)
df.reset_index(drop=True,inplace=True)
df

Name                     Model Transmission  \
0           2020 Renault Kwid  1.0 RXT Opt AT Automatic    Automatic   
1        2013 Maruti Alto 800                LXI Manual       Manual   
2     2020 Maruti New Wagon-R            VXI 1.0 Manual       Manual   
3           2019 Maruti Swift                VXI Manual       Manual   
4          2020 Maruti Baleno      SIGMA 1.2 K12 Manual       Manual   
...                       ...                       ...          ...   
5335  2016 Maruti Wagon R 1.0                      VXI        Manual   
5336         2015 Hyundai Eon                   SPORTZ        Manual   
5337  2012 Maruti Wagon R 1.0                      VXI        Manual   
5338          2018 Tata Tiago          XT 1.2 REVOTRON        Manual   
5339     2017 Maruti Alto 800                      VXI        Manual   

     Km_driven      Owner    Fuel  Discount Discount_price Sale_price  
0       3,252   1st Owner  Petrol   ₹5,000        4,31,699   4,36,699  
1      13,807   1st Owner  Petrol  ₹11,000        2,16,899   2,27,899  
2       3,865   1st Owner  Petrol   ₹5,000        5,16,399   5,21,399  
3       3,362   1st Owner  Petrol   ₹5,000        6,20,699   6,25,699  
4       5,645   1st Owner  Petrol   ₹6,000        5,75,299   5,81,299  
...        ...        ...     ...       ...            ...        ...  
5335   33,463   1st Owner  Petrol   ₹6,000        3,88,699   3,94,699  
5336   15,346   3rd Owner  Petrol  ₹20,000        3,01,299   3,21,299  
5337   38,553   1st Owner  Petrol  ₹12,000        3,15,199   3,27,199  
5338   11,253   1st Owner  Petrol   ₹7,000        4,86,799   4,93,799  
5339   21,239   1st Owner  Petrol  ₹37,000        3,20,999   3,57,999  

[5340 rows x 9 columns]

In [141]:
df.shape

(5340, 9)

Here we can see that we have 5340 rows and 9 columns in our dataset.

### Saving the dataframe in csv

In [144]:
df.to_csv("car.csv")